### DISCUSSION TITLE
Tree Filter and/or Programmatically Tick Nodes

### DISCUSSION THREAD

**Discussion by Rubp-007** [zauberzeug/nicegui#1924] [Q&A] [answered]
### Question

Hi community!

I have a tree with 300+ nodes of 5 levels, so it's a bit complicated to navigate through different nodes. I would like to implement a filter box (https://quasar.dev/vue-components/tree#accordion-filtering-and-selectable) that sits above the tree, then I can type and filter the nodes. How should I do that?

As two alternatives, I'm thinking:
1) Use a single select box that once an option is selected, the node in the tree will be ticked. I've tried to use `tree._props['ticked'].append(node)`, but it didn't trigger the tree's "on_tick" callback.
2) Use a multi-select box that is synced with all the ticked nodes. But still, when I send the selected nodes from the box to the tree, the callback won't be fired :(

BTW, I noticed that in NiceGUI, the node won't change the expand/collapse state when it's selected, but in Quasar it will. Is there any way to make the NiceGUI tree the same as the QTree?

Thanks!

Edit:
If we can programmatically tick the nodes of the Tree, it will be handy :)

**Comment by falkoschindler** [comment, marked as answer]
Here is an example with filter and programmatically ticking a node:
```py
tree = ui.tree([
    {'id': 'A', 'children': [{'id': 'A1'}, {'id': 'A2'}]},
    {'id': 'B', 'children': [{'id': 'B1'}, {'id': 'B2'}]},
], label_key='id', tick_strategy='leaf', on_tick=lambda e: ui.notify(e.value))

def tick(id):
    tree._props['ticked'].append(id)
    tree.update()

ui.button('Tick A1', on_click=lambda: tick('A1'))

ui.input('Filter', on_change=lambda e: tree.props(f'filter={e.value}'))
```

After modifying `_props`, you need to call `update()`. When using `.props()`, the update happens automatically.

**Reply by Rubp-007** [reply]
Fantastic! Thank you :)

**Reply by cramatt** [reply]
Great example - what if we want to the filter to render before the tree? Is there a way in nicegui to assign tree to variable but defer rendering till later in the code?

**Reply by cramatt** [reply]
ah, this seems to work. Not sure if this is proper python/ nicegui :) 

```python
tree = None
def set_filter(filter):
    tree.props(f'filter={filter}')
    tree.expand()

ui.input('Filter', on_change=lambda e: set_filter(e.value))
tree = ui.tree(...)
```

**Reply by falkoschindler** [reply]
@cramatt There's no need for `tree = None`. As long as `tree` is defined when `set_filter` is called, everything is alright.

**Comment by Rubp-007** [comment]
> BTW, I noticed that in NiceGUI, the node won't change the expand/collapse state when it's selected, but in Quasar it will. Is there any way to make the NiceGUI tree the same as the QTree?

@falkoschindler Hi falkoschindler, I would like to change the expanded state of a node when it's "clicked". How should I do that?

In particular, I'm able to "expand on select", but not "collapse on unselect", because unselect with send a event with value = None, and I'm not sure how to find the unselected node.

**Reply by falkoschindler** [reply]
@Rubp-007 That's a tricky question which took me a while to figure out:

It looks like we're disabling the expansion via "click on the label" by setting the "selected" prop in the initializer of `ui.tree`. I don't find it mentioned anywhere in the Quasar documentation. But when removing this prop, you can expand and collapse nodes by clicking on their label.

```py
tree = ui.tree([
    {'id': 'numbers', 'children': [{'id': '1'}, {'id': '2'}]},
    {'id': 'letters', 'children': [{'id': 'A'}, {'id': 'B'}]},
], label_key='id')
del tree._props['selected']
```

I wonder how we can integrate it more nicely into `ui.tree`'s API. Maybe we need a `selection: bool = True` parameter as an opt-out for node selection... 🤔

**Reply by Rubp-007** [reply]
@falkoschindler Thank you for the workaround! It works like a charm :) 

I'm a newbie for frontend stuff so I'm afraid I couldn't come up with a better solution :( But for my use case, once I enable the "tick box" (by setting tick-strategy), the "select" functionality seems a bit useless, so adding an opt-out option as you suggested, or even forcing the tree to have "select or tick, not both" (i.e. delete "selected" property once tick-strategy is set), is good enough for me.

**Reply by falkoschindler** [reply]
I created feature request #2014. Would you like to create a pull request? 🙂

**Reply by Rubp-007** [reply]
@falkoschindler Sure! I'd love to :) I'll create the PR later this week.
